core/src/main/java/com/linecorp/armeria/client/endpoint/WeightRampingUpStrategy.java

```diff
@@ -0,0 +1,490 @@
+                ((ListenableAsyncCloseable) endpointGroup).whenClosed().thenRunAsync(this::close, executor);
+            }
+        }
+
+        /**
+         * Removes the duplicate endpoints in the specified {@code newEndpoints} and returns a new map
+         * that contains unique endpoints.
+         * The value of the map is the {@link Endpoint} whose {@link Endpoint#weight()} is the summed weight of
+         * same {@link Endpoint}s.
+         */
+        private Map<Endpoint, Endpoint> deduplicateEndpoints(List<Endpoint> newEndpoints) {
+            final Map<Endpoint, Endpoint> newEndpointsMap = new HashMap<>(newEndpoints.size());
+
+            // The weight of the same endpoints are summed.
+            newEndpoints.forEach(
+                    newEndpoint -> newEndpointsMap.compute(newEndpoint, (key, v) -> {
+                        if (v == null) {
+                            return newEndpoint;
+                        }
+                        final int weightSum = newEndpoint.weight() + v.weight();
+                        return newEndpoint.withWeight(weightSum);
+                    }));
+            return newEndpointsMap;
+        }
+
+        @Override
+        public Endpoint selectNow(ClientRequestContext ctx) {
+            return endpointSelector.selectEndpoint();
+        }
+
+        @VisibleForTesting
+        WeightedRandomDistributionEndpointSelector endpointSelector() {
+            return endpointSelector;
+        }
+
+        // Only executed by the executor.
+        private void updateEndpoints(List<Endpoint> newEndpoints) {
+            unhandledNewEndpoints = null;
+            if (rampingUpTaskWindowNanos > 0) {
+                // Check whether we can ramp up with the previous ramped up endpoints which are at the last
+                // of the rampingUpEndpointsEntries.
+                if (shouldRampUpWithPreviousRampedUpEntry()) {
+                    final Set<EndpointAndStep> newlyAddedEndpoints = filterOldEndpoints(newEndpoints);
+                    if (!newlyAddedEndpoints.isEmpty()) {
+                        updateWeightAndStep(newlyAddedEndpoints);
+                        endpointsRampingUp.getLast().addEndpoints(newlyAddedEndpoints);
+                    }
+                    buildEndpointSelector();
+                    return;
+                }
+
+                // Check whether we can ramp up with the next scheduled endpointsRampingUpEntry.
+                if (shouldRampUpWithNextScheduledEntry()) {
+                    // unhandledNewEndpoints will be ramped up when updateWeightAndStep() is executed.
+                    unhandledNewEndpoints = newEndpoints;
+                    return;
+                }
+            }
+
+            final Set<EndpointAndStep> newlyAddedEndpoints = filterOldEndpoints(newEndpoints);
+            if (newlyAddedEndpoints.isEmpty()) {
+                // newlyAddedEndpoints is empty which means that settledEndpoints are changed.
```

Comment:
I think this drifted, I don't see `settledEndpoints` anywhere.

Reply:
Oops. 😅